src/Kiota.Builder/Writers/Shell/ShellCodeMethodWriter.cs

```diff
@@ -0,0 +1,339 @@
+                var origParams = codeElement.OriginalMethod?.Parameters ?? codeElement.Parameters;
+                var requestBodyParam = origParams.OfKind(CodeParameterKind.RequestBody);
+                //var queryStringParam = origParams.OfKind(CodeParameterKind.QueryParameter);
+                //var headersParam = origParams.OfKind(CodeParameterKind.Headers);
+                //var optionsParam = origParams.OfKind(CodeParameterKind.Options);
+                var requestParams = new RequestParams(requestBodyParam, null, null, null);
+                WriteCommandBuilderBody(codeElement, requestParams, isVoid, returnType, writer);
+            }
+        }
+
+        protected void WriteCommandBuilderBody(CodeMethod codeElement, RequestParams requestParams, bool isVoid, string returnType, LanguageWriter writer)
+        {
+            var parent = codeElement.Parent as CodeClass;
+            var classMethods = parent.Methods;
+            var uppercaseRegex = new Regex("([A-Z])");
+            var name = codeElement.SimpleName;
+            name = uppercaseRegex.Replace(name, "-$1").TrimStart('-').ToLower();
+
+            if (codeElement.HttpMethod == null)
+            {
+                // Build method
+                // Puts together the BuildXXCommand objects. Needs a nav property name e.g. users
+                // Command("users") -> Command("get")
+                if (String.IsNullOrWhiteSpace(name))
+                {
+                    // BuildCommand function
+                    if (codeElement.OriginalMethod?.MethodKind == CodeMethodKind.ClientConstructor)
+                    {
+                        var commandBuilderMethods = classMethods.Where(m => m.MethodKind == CodeMethodKind.CommandBuilder && m != codeElement);
+                        writer.WriteLine($"var command = new RootCommand();");
+                        foreach (var method in commandBuilderMethods)
+                        {
+                            writer.WriteLine($"command.AddCommand({method.Name}());");
+                        }
+
+                        writer.WriteLine("return command;");
+                    }
+                    else if (codeElement.OriginalIndexer != null)
+                    {
+                        var targetClass = conventions.GetTypeString(codeElement.OriginalIndexer.ReturnType, codeElement);
+                        var builderMethods = (codeElement.OriginalIndexer.ReturnType as CodeType).TypeDefinition.GetChildElements(true).OfType<CodeMethod>().Where(m => m.IsOfKind(CodeMethodKind.CommandBuilder)).ToList();
+                        conventions.AddRequestBuilderBody(parent, targetClass, writer, prefix: "var builder = ", pathParameters: codeElement.Parameters.Where(x => x.IsOfKind(CodeParameterKind.Path)));
+                        writer.WriteLine("return new Command[] { ");
+                        writer.IncreaseIndent();
+                        for (int i = 0; i < builderMethods.Count; i++)
```

Comment:
foreach

Reply:
I usually prefer foreach, but in this case, I'm using the index within the loop. I find a foreach with a counter easy to mess up.

Reply:
unless I'm missing something the only reason why you're using the index is to iterate on the collection, which a foreach would do for you (pun intended) just fine :)